packages/hasura-storage-js/src/utils/upload.ts

```diff
@@ -0,0 +1,118 @@
+import fetch from 'cross-fetch'
+import FormData from 'form-data'
+import { ErrorPayload, StorageUploadResponse } from './types'
+
+export const fetchUpload = async (
+  backendUrl: string,
+  data: FormData,
+  {
+    accessToken,
+    name,
+    fileId,
+    bucketId,
+    adminSecret,
+    onUploadProgress
+  }: {
+    accessToken?: string
+    name?: string
+    fileId?: string
+    bucketId?: string
+    adminSecret?: string
+    onUploadProgress?: (event: { total: number; loaded: number }) => void
+  } = {}
+): Promise<StorageUploadResponse> => {
+  const headers: HeadersInit = {}
+  if (fileId) {
+    headers['x-nhost-file-id'] = fileId
+  }
+  if (bucketId) {
+    headers['x-nhost-bucket-id'] = bucketId
+  }
+  if (name) {
+    headers['x-nhost-file-name'] = name
+  }
+  if (adminSecret) {
+    headers['x-hasura-admin-secret'] = adminSecret
+  }
+  if (accessToken) {
+    headers['Authorization'] = `Bearer ${accessToken}`
+  }
+
+  const url = `${backendUrl}/files`
+  if (typeof XMLHttpRequest === 'undefined') {
+    // * Non-browser environment: XMLHttpRequest is not available
+    try {
+      const response = await fetch(url, {
+        method: 'POST',
+        headers,
+        body: data as any // * https://github.com/form-data/form-data/issues/513
+      })
+
+      if (!response.ok) {
+        const error: ErrorPayload = {
+          status: response.status,
+          message: await response.text(),
+          // * errors from hasura-storage are not codified
+          error: response.statusText
+        }
+        return { error, fileMetadata: null }
+      }
+      const fileMetadata: {
```

Comment:
We already have a type for this, don't we? I think `FileResponse` is the one we should use here. Although, I'm not sure if we should really cast it explicitly considering that this variable immediately assigned to `fileMetadata`.